File: web/ui/src/lib/meta.ts
import { createContext, useContext, useMemo } from "react";
import {
  CLUSTER_SCOPED_KINDS,
  RECONCILABLE_KINDS,
  RESOURCE_KINDS,
  RESTARTABLE_KINDS,
  SCALABLE_KINDS,
  type ClusterMeta,
  type ProviderInfo,
  type ResourceKindMeta,
} from "../api.ts";

// MetaContext carries the /api/v1/meta payload (distributions, provider matrix, component options)
// fetched once at app start. The forms render from it instead of hard-coding the cluster schema.
export const MetaContext = createContext<ClusterMeta | null>(null);

export function useMeta(): ClusterMeta {
  const meta = useContext(MetaContext);
  if (!meta) {
    throw new Error("useMeta must be used within a MetaContext provider");
  }
  return meta;
}

// COMPONENT_LABELS maps a component spec key to its human-facing label. This is presentation only:
// the option lists and defaults come from the server's /meta payload, not from here.
export const COMPONENT_LABELS: Record<string, string> = {
  cni: "CNI",
  csi: "CSI",
  cdi: "CDI",
  metricsServer: "Metrics Server",
  loadBalancer: "Load Balancer",
  certManager: "Cert Manager",
  policyEngine: "Policy Engine",
  gitOpsEngine: "GitOps Engine",
};

// preferredProvider picks the provider to pre-select for a distribution in the create form. Which
// provider is the zero-config choice depends on the surface: the local `ksail open web`/desktop backend
// (recognizable by its per-provider availability report) is Docker-first — Docker is KSail's only
// required local dependency — while the operator (no gating report) runs in-cluster, so the
// Kubernetes (nested) provider is its zero-config choice. Falls back to the first provider the
// server lists. This is a UX default only — the full set of valid providers still comes from the
// /meta matrix.
export function preferredProvider(providers: string[], status?: ProviderInfo[] | null): string {
  const preferred = status ? "Docker" : "Kubernetes";

  return providers.includes(preferred) ? preferred : (providers[0] ?? "");
}

// availableProviders narrows the providers valid for a distribution (from /api/v1/meta) to those the
// backend reports as usable (from /api/v1/config). When status is null/undefined the backend does
// not gate by availability (e.g. the operator), so every valid provider is returned unchanged.
export function availableProviders(
  valid: string[],
  status: ProviderInfo[] | null | undefined,
): string[] {
  if (!status) {
    return valid;
  }
  const usable = new Set(status.filter((provider) => provider.available).map((p) => p.name));
  return valid.filter((name) => usable.has(name));
}

// unavailableProviders returns the reported providers that are not usable, for surfacing why an
// option is missing. Empty when the backend does not gate.
export function unavailableProviders(
  status: ProviderInfo[] | null | undefined,
): ProviderInfo[] {
  return (status ?? []).filter((provider) => !provider.available);
}

// ResourceKindLists are the kind lists the workload browser renders from: the kind-selector entries
// and the per-action membership lists (which kinds support scale/restart/reconcile, and which are
// delete-denied).
export interface ResourceKindLists {
  kinds: readonly string[];
  scalable: readonly string[];
  restartable: readonly string[];
  reconcilable: readonly string[];
  deleteDenied: readonly string[];
}

// kindNames projects the kinds matching a predicate to their names.
function kindNames(kinds: ResourceKindMeta[], match: (kind: ResourceKindMeta) => boolean): string[] {
  return kinds.filter(match).map((kind) => kind.kind);
}

// resourceKindLists derives the workload browser's kind lists from /api/v1/meta's resourceKinds —
// the runtime source of truth, mirroring the backend's allowlist and predicates so adding a
// browsable kind in Go needs no TypeScript change. Backends that predate the field (and a
// not-yet-loaded meta) fall back to the hand-maintained constants in api.ts; that fallback is
// permanent so the SPA keeps working against older backends.
export function resourceKindLists(meta: ClusterMeta | null): ResourceKindLists {
  const kinds = meta?.resourceKinds;
  if (!kinds) {
    return {
      kinds: RESOURCE_KINDS,
      scalable: SCALABLE_KINDS,
      restartable: RESTARTABLE_KINDS,
      reconcilable: RECONCILABLE_KINDS,
      deleteDenied: CLUSTER_SCOPED_KINDS,
    };
  }

  return {
    kinds: kindNames(kinds, (kind) => kind.browsable),
    scalable: kindNames(kinds, (kind) => kind.scalable),
    restartable: kindNames(kinds, (kind) => kind.restartable),
    reconcilable: kindNames(kinds, (kind) => kind.reconcilable),
    deleteDenied: kindNames(kinds, (kind) => !kind.deletable),
  };
}

// useResourceKinds resolves the workload browser's kind lists from the MetaContext. Unlike useMeta it
// tolerates a missing meta (it falls back to the constants), so views never crash on an older backend
// or before the meta fetch resolves.
export function useResourceKinds(): ResourceKindLists {
  const meta = useContext(MetaContext);

  return useMemo(() => resourceKindLists(meta), [meta]);
}

File: pkg/operator/api/meta.go
package api

import (
	"fmt"
	"net/http"

	"github.com/devantler-tech/ksail/v7/pkg/apis/cluster/v1alpha1"
)

// metaResponse is the static cluster-configuration metadata the SPA renders its forms from. It is
// the single runtime source of truth for the distribution list, the distribution→provider matrix,
// the component option lists, and the workload browser's resource-kind allowlist — the SPA no longer
// hard-codes any of them. The form values derive from the v1alpha1 enums and the resource kinds from
// the allowlist in resources.go, so adding a distribution/provider/enum value/browsable kind in Go
// surfaces in the UI with no TypeScript change.
type metaResponse struct {
	Distributions []string            `json:"distributions"`
	Providers     map[string][]string `json:"providers"`
	Components    []componentMeta     `json:"components"`
	// ResourceKinds is the workload browser's resource-kind allowlist with per-kind verb flags, in
	// the order the kind selector presents them. Additive on the wire: older SPAs ignore it, and the
	// SPA treats its absence (an older backend) as "fall back to the hand-maintained constants".
	ResourceKinds []resourceKindMeta `json:"resourceKinds"`
}

// componentMeta describes one component field: its spec key (the JSON field under spec.cluster), the
// valid enum values, and the API default (so the form pre-selects the value equivalent to "unset").
type componentMeta struct {
	Key     string   `json:"key"`
	Values  []string `json:"values"`
	Default string   `json:"default"`
}

// resourceKindMeta describes one entry of the workload browser's resource-kind allowlist: the kind
// name, its scope, and the verbs/affordances the backend supports for it. Every flag derives from
// the allowlist and predicates in resources.go, so the SPA builds its kind selector and action gates
// from this payload instead of hand-mirroring them in TypeScript.
type resourceKindMeta struct {
	Kind         string `json:"kind"`
	Namespaced   bool   `json:"namespaced"`
	Scalable     bool   `json:"scalable"`
	Restartable  bool   `json:"restartable"`
	Reconcilable bool   `json:"reconcilable"`
	Deletable    bool   `json:"deletable"`
	Browsable    bool   `json:"browsable"`
}

// enumValuer is the subset of a v1alpha1 enum type the meta endpoint needs.
type enumValuer interface {
	ValidValues() []string
	Default() any
}

// clusterMeta assembles the metadata payload from the v1alpha1 enums and validation rules.
func clusterMeta() metaResponse {
	distributions := new(v1alpha1.Distribution).ValidValues()

	allProviders := new(v1alpha1.Provider).ValidValues()
	providers := make(map[string][]string, len(distributions))

	for _, dist := range distributions {
		var valid []string

		for _, name := range allProviders {
			provider := v1alpha1.Provider(name)
			// ValidateForDistribution is the operator's own combination gate, so the matrix can never
			// drift from what the reconciler will actually accept.
			if provider.ValidateForDistribution(v1alpha1.Distribution(dist)) == nil {
				valid = append(valid, name)
			}
		}

		providers[dist] = valid
	}

	components := []componentMeta{
		component("cni", new(v1alpha1.CNI)),
		component("csi", new(v1alpha1.CSI)),
		component("cdi", new(v1alpha1.CDI)),
		component("metricsServer", new(v1alpha1.MetricsServer)),
		component("loadBalancer", new(v1alpha1.LoadBalancer)),
		component("certManager", new(v1alpha1.CertManager)),
		component("policyEngine", new(v1alpha1.PolicyEngine)),
		component("gitOpsEngine", new(v1alpha1.GitOpsEngine)),
	}

	return metaResponse{
		Distributions: distributions,
		Providers:     providers,
		Components:    components,
		ResourceKinds: resourceKindsMeta(),
	}
}

// resourceKindsMeta assembles the resource-kind metadata from the allowlist and its predicates, in
// the curated order of resourceKindEntries. Deletability mirrors DeleteResourceWith's rule exactly:
// delete is denied for cluster-scoped kinds, so it equals the kind's scope.
func resourceKindsMeta() []resourceKindMeta {
	entries := resourceKindEntries()
	kinds := make([]resourceKindMeta, 0, len(entries))

	for _, entry := range entries {
		kinds = append(kinds, resourceKindMeta{
			Kind:         entry.name,
			Namespaced:   entry.kind.Namespaced,
			Scalable:     ResourceKindScalable(entry.name),
			Restartable:  ResourceKindRestartable(entry.name),
			Reconcilable: ResourceKindReconcilable(entry.name),
			Deletable:    entry.kind.Namespaced,
			Browsable:    resourceKindBrowsable(entry.kind),
		})
	}

	return kinds
}

func component(key string, enum enumValuer) componentMeta {
	return componentMeta{
		Key:     key,
		Values:  enum.ValidValues(),
		Default: fmt.Sprintf("%v", enum.Default()),
	}
}

// handleMeta serves the static cluster-configuration metadata. It is an open endpoint (like
// /api/v1/config): the payload is non-sensitive enum metadata the SPA needs to render forms.
func (s *Server) handleMeta(writer http.ResponseWriter, _ *http.Request) {
	writeJSON(writer, http.StatusOK, clusterMeta())
}
